src/drivers/magnetometer/lsm303agr/LSM303AGR.cpp

```diff
@@ -315,9 +315,9 @@ int LSM303AGR::collect()
 		const hrt_abstime timestamp_sample = hrt_absolute_time();
 
 		// switch to right hand coordinate system in place
-		float x_raw = read_reg(OUTX_L_REG_M) + (read_reg(OUTX_H_REG_M) << 8);
-		float y_raw = read_reg(OUTY_L_REG_M) + (read_reg(OUTY_H_REG_M) << 8);
-		float z_raw = -(read_reg(OUTZ_L_REG_M) + (read_reg(OUTZ_H_REG_M) << 8));
+		int16_t x_raw = read_reg(OUTX_L_REG_M) + (read_reg(OUTX_H_REG_M) << 8);
+		int16_t y_raw = read_reg(OUTY_L_REG_M) + (read_reg(OUTY_H_REG_M) << 8);
+		int16_t z_raw = -(read_reg(OUTZ_L_REG_M) + (read_reg(OUTZ_H_REG_M) << 8));
```

Comment:
Why would this be different?

Then right below, PX4Magnetometer update takes x, y, z as float.

Reply:
> Then right below, PX4Magnetometer update takes x, y, z as float.

That's true.
In reality, I use this sensor and that's how I've noticed the issue. 
It took me quite some time to understand that the problem is the float instead of int.
This change fixed the bug and I tested it with the lsm303agr magnetometer.

For example, in [hmc5883](https://github.com/PX4/PX4-Autopilot/blob/850d0bc588af79186286652af4c8293daafd2c4c/src/drivers/magnetometer/hmc5883/HMC5883.cpp#L310) you do read the raw values into int first.
I also use hmc5883 and I had no problem like I had with lsm303agr.

Reply:
Ok, so it's 2 bytes that need to be combined as 16 bit signed integer first (2s complement). I'll do a quick pass to see if this has slipped in anywhere else.